### PR TITLE
fix: [CDS-72737]: Add artifact path filter param to getArtifactList

### PR DESCRIPTION
**Pull request by sarthakkasat**
## Describe your changes
Added artifact path filter param to get artifact list with tag regex provided, if the filter is null it will call the original method.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50139)

### CODE REVIEW

**Comment by sarthakkasat** [comment]
trigger messagecheck

**Review by markoharness — DISMISSED** [review]

File: 960-api-services/src/main/java/io/harness/artifactory/ArtifactoryNgServiceImpl.java [hunk 1]
```diff
@@ -52,8 +52,12 @@ public List<BuildDetails> getBuildDetails(
   }
 
   @Override
-  public List<BuildDetails> getArtifactList(
-      ArtifactoryConfigRequest artifactoryConfig, String repositoryName, String artifactPath, int maxVersions) {
+  public List<BuildDetails> getArtifactList(ArtifactoryConfigRequest artifactoryConfig, String repositoryName,
+      String artifactPath, int maxVersions, String artifactPathFilter, String artifactDirectory) {
+    if (EmptyPredicate.isNotEmpty(artifactPathFilter)) {
+      return getLatestArtifactForArtifactPathFilter(
```
Comment (yogesh-chauhan) [inline 960-api-services/src/main/java/io/harness/artifactory/ArtifactoryNgServiceImpl.java:58]: should we call this method -> getArtifactList() itself ?

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50139)* status: 6 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-api-services/src/main/java/io/harness/artifactory/ArtifactoryNgServiceImpl.java` line 58 at r1](https://reviewable.io/reviews/harness/harness-core/50139#gh-1255635866:-NZk1rP27_sl3ACvOaV7:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/ac730b4183ebee19241e35f57be0de2baafb9791/960-api-services/src/main/java/io/harness/artifactory/ArtifactoryNgServiceImpl.java#L58)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

should we call this method -> getArtifactList() itself ? 
</blockquote></details>

Done.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 5 of 6 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50139)* status: 6 file(s) to review. 1 of 2 approvals obtained.

File: 930-delegate-tasks/src/test/java/io/harness/delegate/task/artifacts/artifactory/ArtifactoryArtifactTaskHandlerTest.java [hunk 2]
```diff
@@ -274,8 +274,8 @@ public void testGetBuildsForGeneric() {
 
     doReturn(Lists.newArrayList(buildDetailsInternal))
         .when(artifactoryNgService)
-        .getArtifactList(
-            artifactoryInternalConfig, sourceAttributes.getRepositoryName(), filePath, MAX_NO_OF_TAGS_PER_ARTIFACT);
+        .getArtifactList(artifactoryInternalConfig, sourceAttributes.getRepositoryName(), filePath,
```
Comment (Abhishek-CDC) [inline 930-delegate-tasks/src/test/java/io/harness/delegate/task/artifacts/artifactory/ArtifactoryArtifactTaskHandlerTest.java:277]: update sourceAttribute with a value for artifactPathFilter to consider that case.
Currently it is always null.

File: 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/artifacts/ArtifactDelegateRequestUtils.java [hunk 3]
```diff
@@ -185,7 +185,7 @@ public ArtifactSourceDelegateRequest getArtifactoryArtifactDelegateRequest(Strin
         artifactDirectory = "/";
       }
       return getArtifactoryGenericArtifactDelegateRequest(repositoryName, repositoryFormat, artifactDirectory, null,
-          null, null, artifactoryConnectorDTO, encryptedDataDetails, ArtifactSourceType.ARTIFACTORY_REGISTRY);
+          tagRegex, null, artifactoryConnectorDTO, encryptedDataDetails, ArtifactSourceType.ARTIFACTORY_REGISTRY);
     }
```
Comment (Abhishek-CDC) [inline 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/artifacts/ArtifactDelegateRequestUtils.java:189]: is this tagRegex that we are considering as artifactPathFilter everywhere?

Comment (sarthakkasat) [reply 950-delegate-tasks-beans/src/main/java/io/harness/delegate/task/artifacts/ArtifactDelegateRequestUtils.java:189]: yes

File: 960-api-services/src/main/java/io/harness/artifactory/ArtifactoryNgServiceImpl.java [hunk 4]
```diff
@@ -52,8 +52,11 @@ public List<BuildDetails> getBuildDetails(
   }
 
   @Override
-  public List<BuildDetails> getArtifactList(
-      ArtifactoryConfigRequest artifactoryConfig, String repositoryName, String artifactPath, int maxVersions) {
+  public List<BuildDetails> getArtifactList(ArtifactoryConfigRequest artifactoryConfig, String repositoryName,
+      String artifactPath, int maxVersions, String artifactPathFilter, String artifactDirectory) {
+    if (EmptyPredicate.isNotEmpty(artifactPathFilter)) {
+      return getArtifactList(artifactoryConfig, repositoryName, artifactDirectory, artifactPathFilter, maxVersions);
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifactory/ArtifactoryNgServiceImpl.java:58]: please add test for this conditional method call

File: 960-api-services/src/main/java/io/harness/artifactory/ArtifactoryNgServiceImpl.java [hunk 5]
```diff
@@ -71,8 +74,8 @@ private List<BuildDetails> getLatestArtifactForArtifactPath(ArtifactoryConfigReq
     return buildDetails;
   }
 
-  private List<BuildDetails> getLatestArtifactForArtifactPathFilter(ArtifactoryConfigRequest artifactoryConfig,
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifactory/ArtifactoryNgServiceImpl.java:74]: why we are renaming this method?

**Review by sarthakkasat — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50139)* status: 6 file(s) to review. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-delegate-tasks/src/test/java/io/harness/delegate/task/artifacts/artifactory/ArtifactoryArtifactTaskHandlerTest.java` line 277 at r2](https://reviewable.io/reviews/harness/harness-core/50139#gh-1257913442:-NZzKHVTBO84fgJDEKYQ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/32e783ee18492c23791d8b0dc9eeb895296de03c/930-delegate-tasks/src/test/java/io/harness/delegate/task/artifacts/artifactory/ArtifactoryArtifactTaskHandlerTest.java#L277)):*
<details><summary><i>Previously, Abhishek-CDC (Abhishek) wrote…</i></summary><blockquote>

update sourceAttribute with a value for artifactPathFilter to consider that case.
Currently it is always null.
</blockquote></details>

Done.
___
*[`960-api-services/src/main/java/io/harness/artifactory/ArtifactoryNgServiceImpl.java` line 0 at r2](https://reviewable.io/reviews/harness/harness-core/50139#gh-1257931878:-NZz9ZqI-VDcxBRkzeJA:bra0lo5) ([raw file](https://github.com/harness/harness-core/blob/32e783ee18492c23791d8b0dc9eeb895296de03c/960-api-services/src/main/java/io/harness/artifactory/ArtifactoryNgServiceImpl.java#L0)):*
<details><summary><i>Previously, Abhishek-CDC (Abhishek) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

why we are renaming this method?
</blockquote></details>

This method just returns the list of artifacts, while the previous name suggests that it is fetching the latest build.
___
*[`960-api-services/src/main/java/io/harness/artifactory/ArtifactoryNgServiceImpl.java` line 58 at r2](https://reviewable.io/reviews/harness/harness-core/50139#gh-1257928752:-NZzKFS4Dsp-nlbTXjel:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/32e783ee18492c23791d8b0dc9eeb895296de03c/960-api-services/src/main/java/io/harness/artifactory/ArtifactoryNgServiceImpl.java#L58)):*
<details><summary><i>Previously, Abhishek-CDC (Abhishek) wrote…</i></summary><blockquote>

please add test for this conditional method call
</blockquote></details>

Done.

**Review by Abhishek-CDC — APPROVED** [review]

**Comment by sarthakkasat** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50139)* status: 6 file(s) to review. 4. unresolved discussions.

**Comment by yogesh-chauhan** [comment]
trigger review

**Pull request merged**